numpyro/infer/hmc_gibbs.py

```diff
@@ -156,42 +156,106 @@ def potential_fn(z_gibbs, z_hmc):
         return HMCGibbsState(z, hmc_state, rng_key)
 
 
-def _discrete_step(rng_key, z_discrete, pe, potential_fn, idx, support_size):
+def _discrete_gibbs_propose(rng_key, z_discrete, pe, potential_fn, idx, support_size):
     # idx: current index of `z_discrete_flat` to update
     # support_size: support size of z_discrete at the index idx
 
     # get z proposal
     z_discrete_flat, unravel_fn = ravel_pytree(z_discrete)
+    pe_0 = pe
 
     # XXX: we can't vmap potential_fn over all proposals and sample from the conditional
     # categorical distribution because support_size is a traced value, i.e. its value
     # might change across different discrete variables;
     # so here we will loop over all proposals and use an online scheme to sample from
     # the conditional categorical distribution
     def body_fn(i, val):
-        rng_key, z, pe, weight_logsumexp = val
-        rng_key, rng_accept = random.split(rng_key)
+        rng_key, z, pe, log_weight_sum = val
+        rng_key, rng_transition = random.split(rng_key)
         proposal = jnp.where(i >= z_discrete_flat[idx], i + 1, i)
         z_new_flat = ops.index_update(z_discrete_flat, idx, proposal)
         z_new = unravel_fn(z_new_flat)
         pe_new = potential_fn(z_new)
-        weight_new = pe - pe_new
+        log_weight_new = pe_0 - pe_new
```

Comment:
This is a bug in the previous implementation. We should use the same constant `pe_0` during those transitions.